core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java

```diff
@@ -0,0 +1,297 @@
+            out.writeOptionalString(command.node);
+            out.writeOptionalString(command.allocationId);
+            out.writeBoolean(command.ignoreDeciders);
+        }
+
+        @Override
+        public AllocatePrimaryAllocationCommand fromXContent(XContentParser parser) throws IOException {
+            String index = null;
+            int shardId = -1;
+            String nodeId = null;
+            String allocationId = null;
+            boolean ignoreDeciders = false;
+
+            String currentFieldName = null;
+            XContentParser.Token token;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    currentFieldName = parser.currentName();
+                } else if (token.isValue()) {
+                    if ("index".equals(currentFieldName)) {
+                        index = parser.text();
+                    } else if ("shard".equals(currentFieldName)) {
+                        shardId = parser.intValue();
+                    } else if ("node".equals(currentFieldName)) {
+                        nodeId = parser.text();
+                    } else if ("allocation_id".equals(currentFieldName) || "allocationId".equals(currentFieldName)) {
+                        allocationId = parser.text();
+                    } else if ("ignore_deciders".equals(currentFieldName) || "ignoreDeciders".equals(currentFieldName)) {
+                        ignoreDeciders = parser.booleanValue();
+                    } else {
+                        throw new ElasticsearchParseException("[{}] command does not support field [{}]", NAME, currentFieldName);
+                    }
+                } else {
+                    throw new ElasticsearchParseException("[{}] command does not support complex json tokens [{}]", NAME, token);
+                }
+            }
+            if (index == null) {
+                throw new ElasticsearchParseException("[{}] command missing the index parameter", NAME);
+            }
+            if (shardId == -1) {
+                throw new ElasticsearchParseException("[{}] command missing the shard parameter", NAME);
+            }
+            if (nodeId == null && allocationId == null) {
+                throw new ElasticsearchParseException("[{}] command missing either node or allocation_id parameter", NAME);
+            }
+            return new AllocatePrimaryAllocationCommand(new ShardId(index, shardId), nodeId, allocationId, ignoreDeciders);
+        }
+
+        @Override
+        public void toXContent(AllocatePrimaryAllocationCommand command, XContentBuilder builder, ToXContent.Params params, String objectName) throws IOException {
+            if (objectName == null) {
+                builder.startObject();
+            } else {
+                builder.startObject(objectName);
+            }
+            builder.field("index", command.shardId().index().name());
```

Comment:
Other places we use a `Fields` inner class for the field names. Not here? I'm not really 100% sure why we don't just use strings for the field names so I can't speak with confidence about this.